### ISSUE TITLE
`renv::restore()` doesn't use RemoteSHA recorded in the lockfile when running in Docker container

### ISSUE THREAD

**Issue by yogat3ch** [rstudio/renv#1883]
Hi @kevinushey and renv dev team,
We're using `renv::restore` as part of our Docker build process and just discovered that when `renv::restore` runs while building the Docker image it's not respecting the RemoteSHA recorded in the lockfile for github dependencies, and is rather using the most recent commit. I was under the assumption that `renv::restore` uses the RemoteSHA specified in the renv.lock file, but that doesn't seem to be the case. Am I missing a configuration option to ensure that `renv::restore` uses the RemoteSHA from the lock file?

**Comment by kevinushey** [comment]
Can you please provide a reproducible example? Also note that the field is normally called RemoteSha, in case you're generating your renv lock files by hand.

**Comment by yogat3ch** [comment]
Hi @kevinushey,
The repo is up [here](https://github.com/yogat3ch/testRenvDocker)! 
To reproduce:
1. Clone the repo
2. Set your architecture in the Dockerfile if you don't want arm64
3.  Start docker
4. Run 
```bash
docker build -f Dockerfile -t test:arm64_latest . 2>&1 | tee build.log
```
And check the `build.log`.
You can also just take a look at the _build.log_ file that's present from when I just ran it and see that the hash in the log is different from the hash in the _renv.lock_.
Thank you in advance for looking into this!

**Comment by kevinushey** [comment]
Thanks -- much appreciated! I'll try to take a closer look tomorrow.

**Comment by yogat3ch** [comment]
Sounds good, thank you @kevinushey !

**Comment by kevinushey** [comment]
Thanks! I've confirmed the issue, and it's specific to usages of GitHub repositories when `renv.config.pak.enabled` is set. (To be specific, it's a bug in how `renv` is trying to use `pak`, not `pak`).

We're creating the remote specification using RemoteRef rather than RemoteSha, which is incorrect.

For the curious:

https://github.com/rstudio/renv/blob/5cbf955aef25396ede5d27d5fb375611b4251a4c/R/pak.R#L126-L127

but we prefer using RemoteRef when available when formatting a remote:

https://github.com/rstudio/renv/blob/5cbf955aef25396ede5d27d5fb375611b4251a4c/R/records.R#L131-L138

so we'll want to adjust how we format the remote for use by `pak` to prefer using the RemoteSha instead.

Thanks again for providing the reproducible example; it made it far easier to ascertain the root cause here!

**Comment by yogat3ch** [comment]
@kevinushey Awesome, thank you for diagnosing the issue! So is it just a matter of switching the order of the `if else` statement there?

**Comment by kevinushey** [comment]
Not quite; that function is used both when trying to format a record as a remote for display (where we might want to truncate the SHA since just the first few characters are probably fine), whereas we'd want to preserve the entire SHA when using that for an actual installation request. I'm going to poke at it a bit further.

**Comment by yogat3ch** [comment]
Hi @kevinushey,
Verified working on my end! 
Thank you for fixing that so quickly!

**Comment by yogat3ch** [comment]
Hi @kevinushey,
It looks like this is having some unexpected downstream consequences with pak when multiple entries for a single dependency exist:
```
#19 49.58 Error:
--
2478 | #19 49.58 ! error in pak subprocess
2479 | #19 49.58 Caused by error:
2480 | #19 49.58 ! Could not solve package dependencies:
2481 | #19 49.58 * yogat3ch/DTExtend@436c239d6c7c7c4bc4df1d7ff2acc9b03fb0934a: Can't install dependency yogat3ch/UU (>= 1.44.0)
2482 | #19 49.58 * yogat3ch/UU: Conflicts with yogat3ch/UU@1935fc3262898ed6c93237407565e3b910579616
2483 | #19 49.58 * yogat3ch/OpenMCE@a1ad9b528138a58b0537531aa8dcc00be12e7850: Can't install dependency yogat3ch/UU
2484 | #19 49.58 * yogat3ch/bs4Dash@1d4d4618e47457067fd07e17c840e58f0213c684: Can't install dependency yogat3ch/UU (>= 1.5.1)
2485 | #19 49.58 * Martin-McCoy/DMDUAPI@3846630ba715eef40230ae5e81608dc129bbac32: Can't install dependency yogat3ch/UU (>= 1.39.3)
2486 | #19 49.58 * nlme@3.1-162: dependency conflict
2487 | #19 49.58 * Martin-McCoy/shinyVirga@59764669b9b9fe679220c78130e038fd548d0231: Can't install dependency yogat3ch/UU (>= 1.43.0)
2488 | #19 49.58 * JohnCoene/tippy@0d89b2a5cf67c7ba2868a4777125326ef462ec76: dependency conflict
2489 | #19 49.58 * Martin-McCoy/virgaUtils@3dbd35799aa988f6b43ebe512a64069f77356e2c: Can't install dependency yogat3ch/UU (>= 0.24.2)
2490 | #19 49.58 * codetools@0.2-19: dependency conflict
2491 | #19 49.58 * R6@2.5.1: dependency conflict
2492 | #19 49.58 * cpp11@0.4.7: dependency conflict
2493 | #19 49.58 * credentials@2.0.1: dependency conflict
2494 | #19 49.58 * curl@5.2.1: dependency conflict
2495 | #19 49.58 * dplyr@1.1.4: dependency conflict


```

Does it make sense to you that this would happen or should I try to create a reprex?

**Comment by yogat3ch** [comment]
Hey @kevinushey,
Bumping this as the workaround we tried does not fix the issue. `renv` will use the latest version of every Github package if the package is installed from Github, even if the version number has changed and even if the _DESCRIPTION_ explicitly specifies a package version. 
A quick scan over the recent issues suggests that [9328cfe](https://github.com/rstudio/renv/commit/9328cfef9e4430f30fa38e28b671d2a9b1094891) might be causing other people's builds to fail also if they installed `renv` from Github, given how `renv` is automatically using the most recent version of everything installed from Github.

**Comment by yogat3ch** [comment]
We just built these two utilities for overwriting the latest version that `renv` installs with the version specified in the lockfile. Hopefully they prove helpful for fixing the issue:
```r
renv_gh_pkgs <- function(file = "renv.lock") {
  lf <- if (!inherits(file, "renv_lockfile"))
    renv::lockfile_read(file)
  else
    file
  gh_pkgs <- purrr::keep(lf$Packages, \(.x) .x$Source == "GitHub")
}

renv_gh_pkgs_install <- function(pkgs) {
  purrr::walk(pkgs, \(.x) {
    remotes::install_github(
      sprintf("%s/%s", .x$RemoteUsername, .x$RemoteRepo),
      ref = .x$RemoteSha,
      force = TRUE
    )
  })

}
```

We've implemented after `renv::restore` in the Dockerfile as follows:

```Dockerfile

# Copies the renv.lock file from your local project into the Docker image.
# This lock file ensures that R packages are installed at specific versions for reproducibility.
COPY renv.lock renv.lock

# Copies the .Renviron file, which typically contains environment variables for R.
# This can include API keys or other credentials needed for your R environment.
COPY .Renviron .Renviron

# Copies the .Rprofile, which can contain R commands to run every time an R session starts.
# This can be used to set global options or load certain packages by default.
COPY .Rprofile .Rprofile

# Copies the .Rbuildignore file into the Docker image.
# This file indicates which files and directories should be ignored by R build tools.
COPY .Rbuildignore .Rbuildignore

# Copies the 'renv' directory, which contains R package caches and configuration files.
# This directory helps 'renv' manage packages more efficiently.
COPY renv renv

# Sets the RENV_PATHS_LIBRARY environment variable.
# This tells 'renv' where to store and look for installed R packages.
ENV RENV_PATHS_LIBRARY /usr/local/lib/R/site-library

# Installs the 'remotes' package, which is useful for installing R packages from sources such as GitHub.
RUN R --quiet -e "install.packages('remotes')"

# Installs the 'renv' package, a package management tool that helps isolate project-specific dependencies.
# This version includes a fix that ensures the SHA of Github packages recorded in the lockfile are installed instead of the most recent version.
RUN R --quiet -e "install.packages('renv')"

# Restores the R environment based on the renv.lock file.
# This ensures that all R package dependencies are installed.
RUN R --quiet -e "renv::diagnostics();renv::restore(library = '/usr/local/lib/R/site-library')"

# Copy functions for installing specific versions
COPY R/utils_renv.R R/utils_renv.R

# Overwrite whatever versions the `renv::restore` installed for Github Repositories with those recorded in the lockfile
RUN R --quiet -e "source('R/utils_renv.R'); renv_gh_pkgs_install(renv_gh_pkgs());"

```

**Comment by kevinushey** [comment]
Could you also share the lockfile where you saw this issue, so I can try to reproduce locally and better understand the issue?

**Comment by yogat3ch** [comment]
Hi @kevinushey,
I've updated the [`testRenvDocker` repo](https://github.com/yogat3ch/testRenvDocker) with a lockfile that is similar to the one we're using except that it has private Github repos removed. I've installed the dev version of `renv` there so It will throw the dependency conflict error mentioned above.

We've happened upon another issue, which I can move over to another issue if need be.

We re-arranged our workaround a bit in the Dockerfile:
```Dockerfile
# This tells 'renv' where to store and look for installed R packages.
ENV RENV_PATHS_LIBRARY /usr/local/lib/R/site-library

# Installs the 'remotes' package, which is useful for installing R packages from sources such as GitHub.
RUN R --quiet -e "install.packages('remotes', lib = Sys.getenv('RENV_PATHS_LIBRARY'))"

# Installs the 'renv' package, a package management tool that helps isolate project-specific dependencies.
# This version includes a fix that ensures the SHA of Github packages recorded in the lockfile are installed instead of the most recent version.
RUN R --quiet -e "install.packages('renv', lib = Sys.getenv('RENV_PATHS_LIBRARY'))"

# Copy functions for installing specific versions
COPY R/utils_renv.R R/utils_renv.R

# Need purrr for subsequent calls
RUN R --quiet -e "install.packages('purrr', lib = Sys.getenv('RENV_PATHS_LIBRARY'))"
# Overwrite whatever versions the `renv::restore` installed for Github Repositories with those recorded in the lockfile
RUN R --quiet -e "source('R/utils_renv.R'); renv::deactivate(); renv_gh_pkgs_install(renv_gh_pkgs());"

# Restores the R environment based on the renv.lock file.
# This ensures that all R package dependencies are installed.
RUN R --quiet -e "renv::deactivate();source('R/utils_renv.R'); renv::restore(exclude = names(renv_gh_pkgs()))"


```

In order to install the Github repos first, using their specific RemoteSha values, and then exclude them from the `renv::restore` so as not to install the dependencies twice. 

For some reason, switching the order of operations has now produced this obtuse error when running `renv::restore`:
```
#21 146.7 ℹ Installing system requirements
#21 146.7 ℹ Executing `sh -c apt-get -y update`
#21 150.7 Error: 
#21 150.7 ! error in pak subprocess
#21 150.7 Caused by error in `processx::run(sh, cmdline, stdout_callback = callback, stderr_to_stdout = TRUE)`:
#21 150.7 ! System command 'sh' failed
```
Any idea what's going on here? Seems like it can't run the `sh` command from the background process initialized by `processx::run`? Should I move this to a new issue?

<details>
<summary> Full build.log output with traceback </summary>

```
#21 143.7 ✔ Got parsnip 1.0.4 (source) (971.48 kB)
#21 144.5 ✔ Got plotly 4.10.1 (source) (3.89 MB)
#21 144.5 ✔ Got ggbeeswarm 0.7.2 (source) (1.72 MB)
#21 146.7 ✔ Got vipor 0.4.5 (source) (4.70 MB)
#21 146.7 ℹ Installing system requirements
#21 146.7 ℹ Executing `sh -c apt-get -y update`
#21 150.7 Error: 
#21 150.7 ! error in pak subprocess
#21 150.7 Caused by error in `processx::run(sh, cmdline, stdout_callback = callback, stderr_to_stdout = TRUE)`:
#21 150.7 ! System command 'sh' failed
#21 150.7 ---
#21 150.7 Exit status: 100
#21 150.7 Stdout & stderr:
#21 150.7 Get:1 http://ports.ubuntu.com/ubuntu-ports jammy InRelease [270 kB]
#21 150.7 Err:1 http://ports.ubuntu.com/ubuntu-ports jammy InRelease
#21 150.7   At least one invalid signature was encountered.
#21 150.7 Get:2 http://ports.ubuntu.com/ubuntu-ports jammy-updates InRelease [119 kB]
#21 150.7 Err:2 http://ports.ubuntu.com/ubuntu-ports jammy-updates InRelease
#21 150.7   At least one invalid signature was encountered.
#21 150.7 Get:3 http://ports.ubuntu.com/ubuntu-ports jammy-backports InRelease [109 kB]
#21 150.7 Err:3 http://ports.ubuntu.com/ubuntu-ports jammy-backports InRelease
#21 150.7   At least one invalid signature was encountered.
#21 150.7 Get:4 http://ports.ubuntu.com/ubuntu-ports jammy-security InRelease [110 kB]
#21 150.7 Err:4 http://ports.ubuntu.com/ubuntu-ports jammy-security InRelease
#21 150.7   At least one invalid signature was encountered.
#21 150.7 Reading package lists...
#21 150.7 W: GPG error: http://ports.ubuntu.com/ubuntu-ports jammy InRelease: At least one invalid signature was encountered.
#21 150.7 E: The repository 'http://ports.ubuntu.com/ubuntu-ports jammy InRelease' is not signed.
#21 150.7 W: GPG error: http://ports.ubuntu.com/ubuntu-ports jammy-updates InRelease: At least one invalid signature was encountered.
#21 150.7 E: The repository 'http://ports.ubuntu.com/ubuntu-ports jammy-updates InRelease' is not signed.
#21 150.7 W: GPG error: http://ports.ubuntu.com/ubuntu-ports jammy-backports InRelease: At least one invalid signature was encountered.
#21 150.7 E: The repository 'http://ports.ubuntu.com/ubuntu-ports jammy-backports InRelease' is not signed.
#21 150.7 W: GPG error: http://ports.ubuntu.com/ubuntu-ports jammy-security InRelease: At least one invalid signature was encountered.
#21 150.7 E: The repository 'http://ports.ubuntu.com/ubuntu-ports jammy-security InRelease' is not signed.
#21 150.7 ---
#21 150.7 Backtrace:
#21 150.7 1. renv::restore(exclude = names(renv_gh_pkgs()))
#21 150.7 2. renv:::renv_pak_restore(lockfile = lockfile, packages = packages, exclude = exclu…
#21 150.7 3. pak$pkg_install(remotes)
#21 150.7 4. pak:::remote(function(...) get("pkg_install_do_plan", asNamespace("pak"))(...), …
#21 150.7 5. err$throw(res$error)
#21 150.7 ---
#21 150.7 Subprocess backtrace:
#21 150.7  1. base::withCallingHandlers(cli_message = function(msg) { …
#21 150.7  2. get("pkg_install_do_plan", asNamespace("pak"))(...)
#21 150.7  3. proposal$install_sysreqs()
#21 150.7  4. pkgdepends:::sysreqs_install(cmds, config)
#21 150.7  5. base::lapply(cmds, function(cmd) { …
#21 150.7  6. local FUN(X[[i]], ...)
#21 150.7  7. processx::run(sh, cmdline, stdout_callback = callback, stderr_to_stdout = …
#21 150.7  8. base::throw(new_process_error(res, call = sys.call(), echo = echo, …
#21 150.7  9. | base::signalCondition(cond)
#21 150.7 10. global (function (e) …
#21 150.7 Traceback (most recent calls last):
#21 150.7 6: renv::restore(exclude = names(renv_gh_pkgs()))
#21 150.7 5: renv_pak_restore(lockfile = lockfile, packages = packages, exclude = exclude, 
#21 150.7        project = project)
#21 150.7 4: pak$pkg_install(remotes)
#21 150.7 3: remote(function(...) get("pkg_install_do_plan", asNamespace("pak"))(...), 
#21 150.7        list(proposal = NULL))
#21 150.7 2: err$throw(res$error)
#21 150.7 1: base::stop(cond)
#21 150.7 Execution halted
#21 ERROR: process "/bin/sh -c R --quiet -e \"renv::deactivate();source('R/utils_renv.R'); renv::restore(exclude = names(renv_gh_pkgs()))\"" did not complete successfully: exit code: 1
------
 > [stage-2 16/16] RUN R --quiet -e "renv::deactivate();source('R/utils_renv.R'); renv::restore(exclude = names(renv_gh_pkgs()))":
150.7 Traceback (most recent calls last):
150.7 6: renv::restore(exclude = names(renv_gh_pkgs()))
150.7 5: renv_pak_restore(lockfile = lockfile, packages = packages, exclude = exclude, 
150.7        project = project)
150.7 4: pak$pkg_install(remotes)
150.7 3: remote(function(...) get("pkg_install_do_plan", asNamespace("pak"))(...), 
150.7        list(proposal = NULL))
150.7 2: err$throw(res$error)
150.7 1: base::stop(cond)
150.7 Execution halted
------
Dockerfile:104
--------------------
 102 |     # Restores the R environment based on the renv.lock file.
 103 |     # This ensures that all R package dependencies are installed.
 104 | >>> RUN R --quiet -e "renv::deactivate();source('R/utils_renv.R'); renv::restore(exclude = names(renv_gh_pkgs()))"
 105 |     
 106 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c R --quiet -e \"renv::deactivate();source('R/utils_renv.R'); renv::restore(exclude = names(renv_gh_pkgs()))\"" did not complete successfully: exit code: 1
```

</details>

**Comment by kmasiello** [comment]
Hi @kevinushey , I believe this is related:

I am seeing that some packages in a `renv.lock` are completely ignored by `renv::restore()` when building a Dockerfile. The build logs don't even list the packages in the download/install list. 

The attached lock file generates the following build output:
[renv.lock.txt](https://github.com/user-attachments/files/16503321/renv.lock.txt)
```
#25 0.572 > renv::restore(lockfile='renv.lock')
#25 4.126 The following package(s) will be updated:
#25 4.126 
#25 4.126 # RSPM -----------------------------------------------------------------------
#25 4.126 - cli     [3.6.2 -> 3.6.3]
#25 4.126 - rlang   [1.1.3 -> 1.1.4]
#25 4.126 
#25 4.140 # Downloading packages -------------------------------------------------------
#25 4.223 - Downloading cli from RSPM ...                 OK [1.2 Mb in 1.7s]
#25 6.179 - Downloading rlang from RSPM ...               OK [1.5 Mb in 1.8s]
#25 8.192 Successfully downloaded 2 packages in 4.1 seconds.
#25 8.192 
#25 8.193 # Installing packages --------------------------------------------------------
#25 8.274 - Installing cli ...                            OK [installed binary and cached in 1.3s]
#25 9.783 - Installing rlang ...                          OK [installed binary and cached in 1.4s]
#25 11.39 > 
#25 11.39 > 
#25 DONE 13.4s
```

As you can see, bit, bit64, blob, vctrs, glue, lifecycle, and pkgconfig are all ignored. The lockfile restores properly outside of a Docker context. 
Interestingly, in the larger project where I discovered this issue, I'm also installing `forcats`, which lists `glue` as a dependency, and there are no errors thrown.

**Comment by kevinushey** [comment]
@kmasiello I suspect this is because those packages are already installed in another library on your library paths, or because those packages were already installed via some separate mechanism. Is that possible? Are you able to share the whole build log?

**Comment by kmasiello** [comment]
<_ceases pulling out hair_> @kevinushey I think you're right. They are buried in our base r-session-complete image. Still haven't found exactly where, but running a clean session with just `ghcr.io/rstudio/r-session-complete:ubuntu2204` has those offending packages in the R 4.4 system library.  I retract my comments and shall take a walk to try to recover my sanity. 🚶‍♀️

**Comment by karawoo** [comment]
I think I have encountered the same issue @yogat3ch mentioned in [this comment](https://github.com/rstudio/renv/issues/1883#issuecomment-2111223630). Here is a reprex in case it helps. I think the important thing here is the lockfile contains a github package `ggbarf` which itself depends on another github package `ggtextures`. If the version of ggtextures in the lockfile is not the most recent, restoring with pak fails. 

```sh
mkdir renv-pak-test && cd renv-pak-test
# might need to clear the pak cache first? not sure
# Rscript -e "pak::cache_clean(); pak::meta_clean(force = TRUE)"

# Install packages. ggbarf is a github package that depends on another github
# package, ggtextures. here I'm installing an older commit of ggtextures --
# that's what should make it into the lockfile 
Rscript -e "remotes::install_github('rstudio/renv@5725cdee36c6d0b8c32fef4f0bd4b4e1da9e73e2'); remotes::install_github('clauswilke/ggtextures@156c0c6799b53b317d4dfc5ba284254f0fd68ed5'); remotes::install_github('karawoo/ggbarf@master')"
echo "library('ggbarf')" >> script.R
Rscript -e "renv::init()"

# succeeds
rm -rf renv/library
Rscript -e "renv::restore()"

# fails
rm -rf renv/library
Rscript -e "options('renv.config.pak.enabled'=TRUE); renv::restore()"
```

<details>

```
Rscript -e "options('renv.config.pak.enabled'=TRUE); renv::restore()"
# Bootstrapping renv 1.0.10.9000 [sha: 5725cde] ------------------------------
- Downloading renv ... OK
- Installing renv  ... OK

- One or more packages recorded in the lockfile are not installed.
- Use `renv::status()` for more details.
The following package(s) will be installed:
- pak [0.8.0]
These packages will be installed into "~/Desktop/renv-pak-test/renv/library/macos/R-4.4/aarch64-apple-darwin20".

# Installing packages --------------------------------------------------------
- Installing pak ...                            OK [linked from cache]
Successfully installed 1 package in 2.4 milliseconds.
✔ Loading metadata database ... done
Error:
! error in pak subprocess
Caused by error:
! Could not solve package dependencies:
* karawoo/ggbarf@d42cc6ec343fa9fd37f4308eb0956c012edd342d: Can't install dependency clauswilke/ggtextures
* clauswilke/ggtextures: Conflicts with clauswilke/ggtextures@156c0c6799b53b317d4dfc5ba284254f0fd68ed5
* R6@2.5.1: dependency conflict
* Matrix@1.7-0: dependency conflict
* RColorBrewer@1.1-3: dependency conflict
* Rcpp@1.0.13: dependency conflict
* cli@3.6.3: dependency conflict
* magrittr@2.0.3: dependency conflict
* nlme@3.1-164: dependency conflict
* purrr@1.0.2: dependency conflict
* rlang@1.1.4: dependency conflict
* scales@1.3.0: dependency conflict
* tibble@3.2.1: dependency conflict
* utf8@1.2.4: dependency conflict
* vctrs@0.6.5: dependency conflict
* viridisLite@0.4.2: dependency conflict
* curl@5.2.3: dependency conflict
* fansi@1.0.6: dependency conflict
* colorspace@2.1-1: dependency conflict
* withr@3.0.1: dependency conflict
* farver@2.1.2: dependency conflict
* ggplot2@3.5.1: dependency conflict
* glue@1.8.0: dependency conflict
* gtable@0.3.5: dependency conflict
* lifecycle@1.0.4: dependency conflict
* isoband@0.2.7: dependency conflict
* labeling@0.4.3: dependency conflict
* munsell@0.5.1: dependency conflict
* pillar@1.9.0: dependency conflict
* pkgconfig@2.0.3: dependency conflict
* lattice@0.22-6: dependency conflict
* magick@2.8.5: dependency conflict
* mgcv@1.9-1: dependency conflict
---
Backtrace:
1. renv::restore()
2. renv:::renv_pak_restore(lockfile = lockfile, packages = packages, exclude = exclude, …
3. pak$pkg_install(pkg = remotes, ask = prompt, )
4. pak:::remote(function(...) get("pkg_install_make_plan", asNamespace("pak"))(...), …
5. err$throw(res$error)
---
Subprocess backtrace:
1. base::withCallingHandlers(cli_message = function(msg) { …
2. get("pkg_install_make_plan", asNamespace("pak"))(...)
3. prop$stop_for_solution_error()
4. private$plan$stop_for_solve_error()
5. pkgdepends:::pkgplan_stop_for_solve_error(self, private)
6. base::throw(new_error("Could not solve package dependencies:\n", msg, …
7. | base::signalCondition(cond)
8. global (function (e) …
Traceback (most recent calls last):
6: renv::restore()
5: renv_pak_restore(lockfile = lockfile, packages = packages, exclude = exclude,
       prompt = prompt, project = project)
4: pak$pkg_install(pkg = remotes, ask = prompt, )
3: remote(function(...) get("pkg_install_make_plan", asNamespace("pak"))(...),
       list(pkg = pkg, lib = lib, upgrade = upgrade, ask = ask,
           start = start, dependencies = dependencies, loaded = loaded_packages(lib)))
2: err$throw(res$error)
1: base::stop(cond)
Execution halted
```

</details>

What's odd however, is that other combinations of dev packages seem to be ok? Here is the same example except instead of ggbarf and ggtextures, I use shinyuieditor and bslib.

```sh
mkdir renv-pak-test && cd renv-pak-test

# installing an older commit of bslib
Rscript -e "remotes::install_github('rstudio/renv@5725cdee36c6d0b8c32fef4f0bd4b4e1da9e73e2'); remotes::install_github('rstudio/bslib@e862e0bdd7deafcad7b001b058995308a0b36654'); remotes::install_github('rstudio/shinyuieditor@main')"
echo "library('shinyuieditor')" >> script.R
Rscript -e "renv::init()"

# succeeds
rm -rf renv/library
Rscript -e "renv::restore()"

# also succeeds
rm -rf renv/library
Rscript -e "options('renv.config.pak.enabled'=TRUE); renv::restore()"
```

One thing that's different between the two is in the first case, ggbarf appears before ggtextures in the lockfile, whereas in the second one shinyuieditor appears after bslib. Not sure if the ordering might matter here.

**Comment by kevinushey** [comment]
For reference, `renv` is constructing a call to `pak::pkg_install()` that's effectively of the form:

```
library(pak)

remotes <- c("MASS@7.3-61", "Matrix@1.7-0", "R6@2.5.1", "RColorBrewer@1.1-3",
  "Rcpp@1.0.13", "cli@3.6.3", "colorspace@2.1-1", "curl@5.2.3",
  "fansi@1.0.6", "farver@2.1.2", "karawoo/ggbarf@d42cc6ec343fa9fd37f4308eb0956c012edd342d",
  "ggplot2@3.5.1", "clauswilke/ggtextures@156c0c6799b53b317d4dfc5ba284254f0fd68ed5",
  "glue@1.8.0", "gtable@0.3.5", "isoband@0.2.7", "labeling@0.4.3",
  "lattice@0.22-6", "lifecycle@1.0.4", "magick@2.8.5", "magrittr@2.0.3",
  "mgcv@1.9-1", "munsell@0.5.1", "nlme@3.1-166", "pillar@1.9.0",
  "pkgconfig@2.0.3", "purrr@1.0.2", "rlang@1.1.4", "scales@1.3.0",
  "tibble@3.2.1", "utf8@1.2.4", "vctrs@0.6.5", "viridisLite@0.4.2",
  "withr@3.0.1")

lib <- tempfile("lib-")
dir.create(lib)
pak::pkg_install(remotes, lib = lib, ask = TRUE)
```

I think the issue arises because `ggbarf` expresses this remote dependency:

```
Remotes:
    clauswilke/ggtextures
```

https://github.com/karawoo/ggbarf/blob/d42cc6ec343fa9fd37f4308eb0956c012edd342d/DESCRIPTION#L14-L15

and that implicitly means, for `pak`, that we want to install the latest version of `ggtextures` from GitHub. That conflicts with the explicit request for "clauswilke/ggtextures@156c0c6799b53b317d4dfc5ba284254f0fd68ed5", and so `pak` fails to install the packages.

I think this would need to be handled in `pak` -- there needs to be some conflict resolution, probably to prefer the explicitly-requested version of a package, versus the one declared in a package's Remotes?

Somewhat more minimal example:

```
library(pak)

remotes <- c(
   "karawoo/ggbarf@d42cc6ec343fa9fd37f4308eb0956c012edd342d",
   "clauswilke/ggtextures@156c0c6799b53b317d4dfc5ba284254f0fd68ed5"
)

lib <- tempfile("lib-")
dir.create(lib)
pak::pkg_install(remotes, lib = lib, ask = TRUE)
```

**Comment by kevinushey** [comment]
It's also possible that `renv` should be constructing the installation request for `pak` in an alternative way, though. @gaborcsardi, do you have any advice?

**Comment by gaborcsardi** [comment]
If you are sure that you have all dependencies covered, then try passing `dependencies = FALSE`.

**Comment by kevinushey** [comment]
Awesome; thanks, I'll do that!

**Comment by kevinushey** [comment]
This gets closer, but with the above example `ggbarf` installation fails. Unfortuantely I don't see the actual installation failure.

```
> renv::restore(exclude = "renv")
✔ Loading metadata database ... done

→ Will install 2 packages.
→ All 2 packages (0 B) are cached.
+ ggbarf       0.0.0.9000 👷🏼🔧 (GitHub: d42cc6e)
+ ggtextures   0.0.1      👷🏽🔧 (GitHub: 156c0c6)
ℹ No downloads are needed, 2 pkgs are cached
ℹ Building ggtextures 0.0.1
ℹ Building ggbarf 0.0.0.9000
✖ Failed to build ggbarf 0.0.0.9000 (128ms)
Error:
! error in pak subprocess
Caused by error in `stop_task_build(state, worker)`:
! Failed to build source package ggbarf.
Type .Last.error to see the more details.
Traceback (most recent calls last):
6: renv::restore(exclude = "renv")
5: renv_pak_restore(lockfile = lockfile, packages = packages, exclude = exclude,
       prompt = prompt, project = project)
4: pak$pkg_install(pkg = remotes, ask = prompt, dependencies = FALSE)
3: remote(function(...) get("pkg_install_do_plan", asNamespace("pak"))(...),
       list(proposal = NULL))
2: err$throw(res$error)
1: base::stop(cond)
> .Last.error
<callr_error/rlib_error_3_0/rlib_error/error>
Error:
! error in pak subprocess
Caused by error in `stop_task_build(state, worker)`:
! Failed to build source package ggbarf.
---
Backtrace:
1. renv::restore(exclude = "renv")
2. renv:::renv_pak_restore(lockfile = lockfile, packages = packages, exclude = exclude, …
3. pak$pkg_install(pkg = remotes, ask = prompt, dependencies = FALSE)
4. pak:::remote(function(...) get("pkg_install_do_plan", asNamespace("pak"))(...), …
5. err$throw(res$error)
---
Subprocess backtrace:
 1. base::withCallingHandlers(cli_message = function(msg) { …
 2. get("pkg_install_do_plan", asNamespace("pak"))(...)
 3. proposal$install()
 4. pkgdepends::install_package_plan(plan, lib = private$library, num_workers = nw, …
 5. base::withCallingHandlers({ …
 6. pkgdepends:::handle_events(state, events)
 7. pkgdepends:::handle_event(state, i)
 8. pkgdepends:::stop_task(state, worker)
 9. pkgdepends:::stop_task_build(state, worker)
10. base::throw(pkg_error("Failed to build source package {.pkg {pkg}}.", …
11. | base::signalCondition(cond)
12. global (function (e) …
```

**Comment by kevinushey** [comment]
I wonder if this is because `pak` still tries to install the packages in parallel at the same time, even though ggtextures is a dependency of ggbarf? Here's a reproducible example.

```r
# work in new library
templib <- tempfile("library-")
dir.create(templib)
.libPaths(templib)

# get pak
install.packages("pak", repos = sprintf("https://r-lib.github.io/p/pak/stable/%s/%s/%s", .Platform$pkgType, R.Version()$os, R.Version()$arch))

# install ggbarf to prepare library
pak::pkg_install("karawoo/ggbarf")

# now, try installing those specific remote requirements
remotes <- c(
  ggbarf = "karawoo/ggbarf@d42cc6ec343fa9fd37f4308eb0956c012edd342d",
  ggtextures = "clauswilke/ggtextures@156c0c6799b53b317d4dfc5ba284254f0fd68ed5"
)

# remove cached packages + installed packages for testing
pak::cache_delete(package = "ggbarf")
pak::cache_delete(package = "ggtextures")
remove.packages(c("ggbarf", "ggtextures"))

# try to install
pak::pkg_install(remotes, dependencies = FALSE)
```

That gives me:

```r
> pak::pkg_install(remotes, dependencies = FALSE)
                                                                                                     
→ Will install 2 packages.
→ Will download 2 packages with unknown size.
+ ggbarf       0.0.0.9000 👷🏻‍♂️🔧 ⬇ (GitHub: d42cc6e)
+ ggtextures   0.0.1      👷🏼‍♀️🔧 ⬇ (GitHub: 156c0c6)
ℹ Getting 2 pkgs with unknown sizes
✔ Got ggbarf 0.0.0.9000 (source) (58.29 kB)                              
✔ Got ggtextures 0.0.1 (source) (1.72 MB)                               
✔ Downloaded 2 packages (1.78 MB) in 631ms                    
ℹ Packaging ggbarf 0.0.0.9000
ℹ Packaging ggtextures 0.0.1
✔ Packaged ggbarf 0.0.0.9000 (379ms)                                      
ℹ Building ggbarf 0.0.0.9000                                              
✔ Packaged ggtextures 0.0.1 (462ms)                                       
ℹ Building ggtextures 0.0.1                                               
✖ Failed to build ggbarf 0.0.0.9000 (147ms)                               
Error:                                                                    
! error in pak subprocess
Caused by error in `stop_task_build(state, worker)`:
! Failed to build source package ggbarf.
Type .Last.error to see the more details.
```

**Comment by gaborcsardi** [comment]
Oh, yeah, that's a good point, then `dependencies = FALSE` is not going to work....
